### PR TITLE
Fix never worked fsync_part_directory/fsync_after_insert/in_memory_parts_insert_sync

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix never worked `fsync_part_directory`/`fsync_after_insert`/`in_memory_parts_insert_sync`

Changelog:
- Fix fsync_part_directory for parts renames
- Fix fsync_after_insert by not closing fd from finalize() (refs #8356, cc @Alex-Burmak )
- Add support of F_FULLSYNC on osx
- Fix fsync_part_directory (missing O_DIRECTORY)
- Smoke coverage of various fsync capabilities

Refs: #11948 (cc @CurtizJ )

*P.S. I've also have pending patches for fsync support for Distributed engine, by they require this patch set.*

### CODE REVIEW

File: src/Common/FileSyncGuard.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+/// OSX does not have O_DIRECTORY
+#ifndef O_DIRECTORY
+#define O_DIRECTORY O_RDWR
+#endif
+
+namespace DB
+{
+
+FileSyncGuard::FileSyncGuard(const DiskPtr & disk_, const String & path)
```
Comment (alexey-milovidov) [inline src/Common/FileSyncGuard.cpp:14]: The name of `FileSyncGuard` class does not reflect the fact that it works only on directories.

**Review by alexey-milovidov — COMMENTED** [review]
LGTM but need to rename the class (and file).

**Comment by alexey-milovidov** [comment]
> P.S. I've also have pending patches for fsync support for Distributed engine, by they require this patch set.

What are the typical reasons of using async inserts to Distributed tables over sync inserts?

**Comment by azat** [comment]
>What are the typical reasons of using async inserts to Distributed tables over sync inserts?

One of the main advantage is batching, asynchronous inserts into Distributed allows batching, so you get something like Buffer table but on disk.
Also it allows to not fail INSERT if remote node is unavailable, files will be just stored on the local filesystem and will be send once the node will back online.
Plus sometimes remote nodes may be overloaded and process INSERT slower, and in this case INSERT into Distributed will not slow anything, but sometimes this is also a disadvantage, since in case of INSERT into underlying table are always slower, then the lag will indefinitely grow.

**Comment by azat** [comment]
>Yandex synchronization check (only for Yandex employees)
>Yandex third-party checks (only for Yandex employees)

Hope that it is missing file in the `ya.make`

And seems that style check is broken - https://clickhouse-test-reports.s3.yandex.net/18845/798fb4e842ed88bfe788c5de9ccac9d9986586d3/style_check/test_run.txt.out.log (and I guess this part is not in the public repository)

**Comment by alexey-milovidov** [comment]
Yes, this is the part that @qoega missed in #18463 (git repository is not present in new check).

**Comment by qoega** [comment]
git repository is present. There is some issue with ignored paths. Contribs should not be checked

**Review by alexey-milovidov — APPROVED** [review]

**Comment by filimonov** [comment]
> > P.S. I've also have pending patches for fsync support for Distributed engine, by they require this patch set.
> 
> What are the typical reasons of using async inserts to Distributed tables over sync inserts?

Proper retries even if first attempt failed.

**Pull request merged**